crates/telio-traversal/src/endpoint_providers/upnp/mod.rs

```diff
@@ -547,24 +573,38 @@ impl<Wg: WireGuard, I: UpnpEpCommands, E: Backoff> Runtime for State<Wg, I, E> {
     type Err = ();
 
     #[allow(index_access_check)]
-    async fn wait_with_update<F>(&mut self, update: F) -> std::result::Result<(), Self::Err>
+    async fn wait_with_update<F>(&mut self, updated: F) -> std::result::Result<(), Self::Err>
     where
         F: Future<Output = BoxAction<Self, std::result::Result<(), Self::Err>>> + Send,
     {
+        pin!(updated);
+
+        if self.is_endpoint_provider_paused {
+            tokio::select! {
+                _ = pending() => {},
+                update = &mut updated => {
+                    return update(self).await;
+                }
+            }
+            return Ok(());
+        }
+
         tokio::select! {
             Ok((len, addr)) = self.udp_socket.recv_from(&mut self.rx_buff) => {
                 let buff = self.rx_buff.clone();
                 let _ = self.handle_ping_rx(&buff[..len], &addr).await;
             }
             _ = &mut self.upnp_interval => {
-                if self.check_endpoint_candidate().await.is_ok(){
-                    self.upnp_interval =  PinnedSleep::new(self.exponential_backoff.get_backoff(), ());
-                } else {
-                    self.exponential_backoff.next_backoff();
-                };
+                if !self.is_endpoint_provider_paused {
```

Comment:
This branch is technically unreachabpe here due to the above if now. While functionally this is correct, I am thinking, maybe we should remove the `if !self.is_endpoint_provider_paused` if to make the select logic slightly simpler?

Reply:
right I forgot the state is replaced and not updated right away